File: src/main/java/com/gsdd/course/app/controller/CourseController.java
package com.gsdd.course.app.controller;

import com.gsdd.course.app.converter.GenericConverter;
import com.gsdd.course.app.model.CourseModel;
import com.gsdd.course.app.persistence.entities.Course;
import com.gsdd.course.app.repository.CourseRepository;
import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cloud.context.config.annotation.RefreshScope;
import org.springframework.data.domain.Sort;
import org.springframework.hateoas.CollectionModel;
import org.springframework.hateoas.Link;
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RequiredArgsConstructor
@RefreshScope
@RestController
@RequestMapping("courses")
public class CourseController {

  @Value("${service.welcome.message}")
  private String welcomeMessage;

  private final CourseRepository courseRepository;
  private final GenericConverter<Course, CourseModel> courseConverter;

  @GetMapping("/welcome")
  public String getWelcomeMsg() {
    return welcomeMessage;
  }

  private CourseModel defineModelWithLinks(Course entity) {
    CourseModel model = courseConverter.convertToDomain(entity);
    Link link =
        WebMvcLinkBuilder.linkTo(
                WebMvcLinkBuilder.methodOn(CourseController.class).getById(model.getCourseId()))
            .withSelfRel();
    model.add(link);
    return model;
  }

  @GetMapping
  public ResponseEntity<CollectionModel<CourseModel>> getAll() {
    List<CourseModel> courses =
        courseRepository.findAll(Sort.by("courseId")).stream()
            .map(this::defineModelWithLinks)
            .toList();
    Link link = WebMvcLinkBuilder.linkTo(CourseController.class).withSelfRel();
    CollectionModel<CourseModel> result = CollectionModel.of(courses, link);
    return ResponseEntity.ok(result);
  }

  @GetMapping("{courseId}")
  public ResponseEntity<CourseModel> getById(@PathVariable("courseId") String courseId) {
    return courseRepository
        .findById(courseId)
        .map(this::defineModelWithLinks)
        .map(ResponseEntity::ok)
        .orElseGet(ResponseEntity.notFound()::build);
  }

  @PostMapping
  public CourseModel saveCourse(@Valid @RequestBody CourseModel course) {
    return defineModelWithLinks(
        courseRepository.saveAndFlush(courseConverter.convertToEntity(course)));
  }

  @PutMapping("{courseId}")
  public ResponseEntity<CourseModel> updateCourse(
      @PathVariable("courseId") String courseId, @Valid @RequestBody CourseModel course) {
    return courseRepository
        .findById(courseId)
        .map(oldCourse -> courseRepository.saveAndFlush(courseConverter.convertToEntity(course)))
        .map(this::defineModelWithLinks)
        .map(ResponseEntity::ok)
        .orElseGet(ResponseEntity.notFound()::build);
  }

  @DeleteMapping("{courseId}")
  public ResponseEntity<Object> deleteCourse(@PathVariable("courseId") String courseId) {
    return courseRepository
        .findById(courseId)
        .map(
            (Course course) -> {
              courseRepository.delete(course);
              return ResponseEntity.noContent().build();
            })
        .orElseGet(ResponseEntity.notFound()::build);
  }
}

File: src/main/kotlin/com/gsdd/course/app/controller/CourseController.kt
package com.gsdd.course.app.controller

import com.gsdd.course.app.converter.GenericConverter
import com.gsdd.course.app.model.CourseModel
import com.gsdd.course.app.persistence.entities.Course
import com.gsdd.course.app.repository.CourseRepository
import jakarta.validation.Valid
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.cloud.context.config.annotation.RefreshScope
import org.springframework.data.domain.Sort
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RefreshScope
@RestController
@RequestMapping("courses")
open class CourseController @Autowired constructor(
    private val courseRepository: CourseRepository,
    private val courseConverter: GenericConverter<Course, CourseModel>,
    @Value("\${service.welcome.message}")
    private val welcomeMessage: String) {

    private fun defineModel(entity: Course): CourseModel =
        courseConverter.convertToDomain(entity)

    @GetMapping("/welcome")
    fun getWelcomeMsg(): String = welcomeMessage

    @GetMapping
    fun getAll(): ResponseEntity<List<CourseModel>> {
        val courses = courseRepository.findAll(Sort.by("courseId"))
            .map { defineModel(it) }
            .toList()
        return ResponseEntity.ok(courses)
    }

    @GetMapping("{courseId}")
    fun getById(@PathVariable("courseId") courseId: String): ResponseEntity<CourseModel?> {
        return courseRepository
            .findById(courseId)
            .map { defineModel(it) }
            .map { ResponseEntity.ok(it) }
            .orElseGet { ResponseEntity.notFound().build() }
    }

    @PostMapping
    fun saveCourse(@RequestBody course: @Valid CourseModel): CourseModel {
        val entity = courseConverter.convertToEntity(course)
        return defineModel(courseRepository.saveAndFlush(entity))
    }

    @PutMapping("{courseId}")
    fun updateCourse(
        @PathVariable("courseId") courseId: String, @RequestBody course: @Valid CourseModel
    ): ResponseEntity<CourseModel> {
        val optionalCourse = courseRepository.findById(courseId)

        return if (optionalCourse.isPresent) {
            val entity = courseConverter.convertToEntity(course)
            val updatedEntity = courseRepository.saveAndFlush(entity)
            ResponseEntity.ok(defineModel(updatedEntity))
        } else {
            ResponseEntity.notFound().build()
        }
    }

    @DeleteMapping("{courseId}")
    fun deleteCourse(@PathVariable("courseId") courseId: String): ResponseEntity<Any> {
        val optionalCourse = courseRepository.findById(courseId)

        return if (optionalCourse.isPresent) {
            courseRepository.delete(optionalCourse.get())
            ResponseEntity.noContent().build()
        } else {
            ResponseEntity.notFound().build()
        }
    }
}
